client/src/components/dataset/sidebars/Filter.jsx

```diff
@@ -18,42 +18,42 @@ export default class Filter extends Component {
 
   componentWillMount() {
     this.setState({
-      op: 'core/filter-column',
-      args: {
-        columnName: this.props.column.columnName,
-        expression: { is: '' },
-      },
-      onError: 'fail',
+      transformation: Immutable.fromJS({
+        op: 'core/filter-column',
+        args: {
+          columnName: this.props.column.get('columnName'),
+          expression: { is: '' },
+        },
+        onError: 'fail',
+      }),
     });
   }
 
   handleChangeExpressionValue(expressionValue) {
-    const { args } = this.state;
-    const [currentExpressionOperator] = getExpressionOperatorAndValue(args.expression);
-    const newArgs = Object.assign({}, args, {
-      expression: {
-        [currentExpressionOperator]: expressionValue,
-      },
+    const { transformation } = this.state;
+    const expression = transformation.getIn(['args', 'expression']);
+    const expressionOperator = expression.keySeq().first();
+    this.setState({
+      transformation: transformation.setIn(
+        ['args', 'expression', expressionOperator], expressionValue
+      ),
     });
-    this.setState(Object.assign({}, this.state, { args: newArgs }));
   }
 
   handleChangeExpressionOperator(expressionOperator) {
-    const { args } = this.state;
-    const currentValue = getExpressionOperatorAndValue(args.expression)[1];
-    const newArgs = Object.assign({}, args, {
-      expression: {
-        [expressionOperator]: currentValue,
-      },
+    const { transformation } = this.state;
+    const expressionValue = transformation.getIn(['args', 'expression', expressionOperator]);
+    this.setState({
+      transformation: transformation.setIn(
+        ['args', 'expression'], Immutable.Map({ [expressionOperator]: expressionValue })),
     });
-    this.setState(Object.assign({}, this.state, { args: newArgs }));
   }
 
   render() {
     const { onClose, onApply, column } = this.props;
-    const { args } = this.state;
+    const { transformation } = this.state;
     const [expressionOperator, expressionValue] =
-      getExpressionOperatorAndValue(args.expression);
+      getExpressionOperatorAndValue(transformation.getIn(['args', 'expression']));
```

Comment:
This is perhaps a very small point, but `expressionValue` can now be `undefined` sometimes.

Because we feed this into the text input as its `value` on line 98, when this is `undefined` it triggers a react warning that we are switching the text input from controlled to uncontrolled

![image](https://cloud.githubusercontent.com/assets/2589977/16652494/e5d9f220-4442-11e6-8892-d4d344d13053.png)

You can reproduce this by opening the Filter sidebar in the dataset editor and switching from "value is" to "value contains"

I don't know if it's intended that `expressionValue` can be undefined sometimes? If it is, you can suppress this warning simply by giving the text input a blank string when there is no `expressionValue` like `<input value={expressionValue || ''}`. If `expressionValue` shouldn't be undefined in the first place, I guess this is a bigger bug